src/splinefunctions.cpp

```diff
@@ -0,0 +1,1083 @@
+    if (get_equidistant_spacing()) {
+        realtype hx2 = 2 * (nodes_[1] - nodes_[0]);
+        for (int i_node = 1; i_node < n_nodes() - 1; i_node++)
+          DSLOPESDP(i_node) =
+            (DVALUESDP(i_node + 1) - DVALUESDP(i_node - 1)) / hx2;
+    } else {
+        for (int i_node = 1; i_node < n_nodes() - 1; i_node++) {
+          realtype dleft =
+            (DVALUESDP(i_node) - DVALUESDP(i_node - 1)) /
+            (nodes_[i_node] - nodes_[i_node - 1]);
+          realtype dright =
+            (DVALUESDP(i_node + 1) - DVALUESDP(i_node)) /
+            (nodes_[i_node + 1] - nodes_[i_node]);
+          DSLOPESDP(i_node) = (dleft + dright) / 2;
+        }
+    }
+
+    // Right boundary (last nodes)
+    switch (last_node_bc_) {
+        case SplineBoundaryCondition::given:
+            DSLOPESDP(last) =
+                  (DVALUESDP(last) - DVALUESDP(last - 1)) /
+                  (nodes_[last] - nodes_[last - 1]);
+            break;
+
+        case SplineBoundaryCondition::zeroDerivative:
+            DSLOPESDP(last) = 0;
+            break;
+
+        case SplineBoundaryCondition::natural:
+            throw AmiException(
+              "Natural boundary condition for Hermite "
+              "splines is not implemented yet."
+            );
+
+        case SplineBoundaryCondition::naturalZeroDerivative:
+            throw AmiException(
+              "Natural boundary condition with zero "
+              "derivative is not allowed for Hermite splines.");
+
+        case SplineBoundaryCondition::periodic:
+            // if one bc is periodic, the other is periodic too
+            DSLOPESDP(last) = DSLOPESDP(0);
+            break;
+
+        default:
+            throw AmiException("Unexpected value for boundary condition.");
+    }
+}
+
+#undef DVALUESDP
+#undef DSLOPESDP
+
+void
+HermiteSpline::compute_coefficients_extrapolation_sensi(
+  int nplist,
+  int spline_offset,
+  gsl::span<realtype> dvaluesdp,
+  gsl::span<realtype> dslopesdp)
+{
+
+    /* Do we want to extrapolate at all? */
+    bool needExtrapolationCoefficients =
+      first_node_ep_ == SplineExtrapolation::constant ||
+      first_node_ep_ == SplineExtrapolation::linear ||
+      last_node_ep_ == SplineExtrapolation::constant ||
+      last_node_ep_ == SplineExtrapolation::linear;
+    if (!needExtrapolationCoefficients)
+        return;
+
+    /* Beyond the spline nodes, we need to extrapolate using a * t + b.
+     * Those coefficients are stored as
+     * [
+     *   D[b_first, p0], D[a_first, p0], D[b_last, p0], D[a_last, p0],
+     *   D[b_first, p1], ... D[a_last, p{nplist-1}]
+     * ]
+     */
+    coefficients_extrapolate_sensi.resize(4 * nplist, 0.0);
+
+    realtype sm0;
+    for (int ip = 0; ip < nplist; ip++) {
+        realtype sp0 = dvaluesdp[spline_offset + ip];
+        switch (first_node_ep_) {
+            /* This whole switch-case-if-else-if-thing could be moved
+             * outside the loop, I know. Yet, it's at most some thousand
```

Comment:
thousand what?

Reply:
Also it looks like there is a fair bit of duplication between the code for first/last nodes. Also I remember that @paulstapor had some issues with spline performance, so it might be worthwhile going the extra mile to have a proper implementation. should be possible to have a rather concise implementation using function pointers, right?

Reply:
above only applies if this function is evaluated more than once per simulation run, otherwise it would be nevertheless be great if duplication between first/last could be reduced.

Reply:
It should be only called once per simulation, the coefficients that are computed there depend only on the parameters and stay constant during the simulation.

Reply:
About reducing duplication between fist/last node, it could be done for that function, but I don't think it would be more readable. For the more complex bc function, where reducing duplication would I think make the code definitely less readable.